agent/pkg/cruntimes/crutils.go

```diff
@@ -0,0 +1,178 @@
+package cruntimes
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io"
+	"net/url"
+	"os/exec"
+	"path"
+	"regexp"
+	"strings"
+
+	"github.com/docker/docker/pkg/stdcopy"
+	"github.com/sirupsen/logrus"
+
+	"github.com/determined-ai/determined/agent/pkg/docker"
+	"github.com/determined-ai/determined/agent/pkg/events"
+	"github.com/determined-ai/determined/master/pkg/model"
+	"github.com/determined-ai/determined/master/pkg/syncx/waitgroupx"
+)
+
+var logLevel = regexp.MustCompile("(?P<level>INFO|WARN|ERROR|FATAL):    (?P<log>.*)")
+
+// ShipCmdLogs does what you might expect from its name.
+func ShipCmdLogs(
+	ctx context.Context,
+	r io.ReadCloser,
+	stdtype stdcopy.StdType,
+	p events.Publisher[docker.Event],
+) {
+	for scan := bufio.NewScanner(r); scan.Scan(); {
+		line := scan.Text()
+		if len(strings.TrimSpace(line)) == 0 {
+			continue
+		}
+
+		var level, log string
+		if matches := logLevel.FindStringSubmatch(line); len(matches) == 3 {
+			level, log = matches[1], matches[2]
+		} else {
+			level, log = model.LogLevelInfo, line
+		}
+
+		if err := p.Publish(ctx, docker.NewTypedLogEvent(level, log, stdtype)); err != nil {
+			logrus.WithError(err).Trace("log stream terminated")
+			return
+		}
+	}
+}
+
+// PprintCommand provides pretty printing of the given command to trace & log outputs.
+func PprintCommand(
+	ctx context.Context,
+	command string,
+	args []string,
+	p events.Publisher[docker.Event],
+	log *logrus.Entry,
+) error {
+	toPrint := command
+	for _, arg := range args {
+		if strings.HasPrefix(arg, "--") { // print each arg on a new line
+			toPrint += " \\\n"
+			toPrint += "\t"
+			toPrint += arg
+		} else {
+			toPrint += " "
+			toPrint += arg
+		}
+	}
+
+	log.Trace(toPrint)
+	if err := p.Publish(ctx, docker.NewLogEvent(
+		model.LogLevelDebug,
+		toPrint,
+	)); err != nil {
+		return err
+	}
+	return nil
+}
+
+// CanonicalizeImage returns the canonicalized image name.
+func CanonicalizeImage(image string) string {
+	url, err := url.Parse(image)
+	isURIForm := err == nil
+	isFSForm := path.IsAbs(image)
+	if isFSForm || (isURIForm && url.Scheme != "") {
+		return image
+	}
+	return fmt.Sprintf("docker://%s", image)
+}
+
+// PullImage implements code sharing for singularity & podman.
+func PullImage(
+	ctx context.Context,
+	req docker.PullImage,
+	p events.Publisher[docker.Event],
+	wg waitgroupx.Group,
```

Comment:
i think we should pass concurrency primitives by pointer not value just incase there is stuff that shouldn't be copied (in this case i think we are ok, but it is subtle). by nature they should be shared.